Orange/widgets/visualize/owmosaic.py

```diff
@@ -410,13 +440,29 @@ def showEvent(self, ev):
         OWWidget.showEvent(self, ev)
         self.update_graph()
 
+    def change_domain(self):
+        if self.data is None or len(self.data) < 2 or len(self.data.domain.attributes) < 1:
+            return
+        _, variable_color = self.get_attr_list()
+        if variable_color is None:
+            color_var = None
+        else:
+            color_var = self.data.domain[variable_color]
+        class_vars = color_var
+        attributes = [v for v in self.data.domain if v != color_var]
+        metas = [v for v in self.data.domain.metas if v != color_var]
+        domain = Domain(attributes, class_vars, metas)
+        data = self.data.from_table(domain, self.data)
+        self.set_data(data)
```

Comment:
Do I see correctly that the combo calls `update_colors`, which calls `change_domain`, which in turn calls `set_data`? And `set_data` is a callback for signal Data, which ... basically resets the widget, including closing the context and so forth?

I think widgets should never call signal handlers like `set_data`. If `set_data` does some things that need to be done when changing the combo (like stopping the vizrank,  updating the graph), extract those into a separate function and call it from `set_data` and from `update_colors` (or `change_domain`).

*Besides*, remove `change_domain` altogether. The widget should output the same domain. Changing the coloring must not replace the class attribute, it should only affect the visual appearance.